### ISSUE TITLE
Writing torrents to Fat, Fat32 or Exfat volumes can create error

### ISSUE THREAD

**Issue by thawkins** [dchang0/torrentwatch-xa#15]
If the torrent name or filename contains characters such as ':' that are illegal in FAT derivative filesystems, the torrent path passed to transmission can result in an error and a failed download.

Optional strict filename character filter should possibly be implemented. 

May Also be a problem on NTFS and HPFS, although I have never observed it there. 

I spotted this when I moved the torrent download directory from an internal linux ext4 volume to an external exfat volume because of space constraints. I will add a small ext4 volume as a workaround, but it would be useful to have the constraint.

**Comment by dchang0** [comment]
Thanks for reporting this.

I can add a filename sanitizer feature. Not sure if it will break the torrent file (it shouldn't), but we can definitely rename the .torrent file.

**Comment by thawkins** [comment]
> Thanks for reporting this.
> 
> I can add a filename sanitizer feature. Not sure if it will break the torrent file (it shouldn't), but we can definitely rename the .torrent file.

'[' and ']' also have the same issue. 

I'm a senior php programer of 15 years, working in linux,. I would be interested in helping out with the project. 

One alternative is just to hash the filenames and use the hashed name.  It's also possible that filename length is an issue on fat derived file names.

**Comment by dchang0** [comment]
Thanks--any help is welcome.
I figure we want to keep the usernames human-readable and avoid hashes. If we DO go with hashes, we could probably just use the torrent hash.

I think what we should do is replace any characters that are illegal in any commonly-used filesystem (macOS, Windows, and LINUX) with minuses and reduce more than one consecutive minus sign to a single minus.

And as you suggest, maybe we truncate the filename length, but I figure we can do that later, after someone reports an issue.
The big challenge will probably be Unicode multibyte characters, but I think those aren't illegal/reserved characters in filesystems because they are multibyte.

Basically, we just need a simple PHP function where we feed in a torrent or magnet filename (includes extension, which should either be .torrent or .magnet), and then returns a "safe" filename with the same extension.

If you feel like writing a function, please do. Otherwise, I'll write one--it's not that hard. Really, the hard part is figuring out which characters are illegal in which filesystems, but that's just a little bit of search engine work.

**Comment by dchang0** [comment]
Come to think of it, there is already a function in torrentwatch-xa that sanitizes filenames. So I just need to apply it to this situation and maybe add the colon and some other illegal characters to the list. This will be easy to fix.

**Comment by dchang0** [comment]
We lucked out. It's a one-line change.

Please try this and test it for your situation.

In the file /var/lib/torrentwatch-xa/lib/twxa_tools.php, please change line 66 to this:

`return preg_replace("/\:|\?|\/|\\|\+|\=|\>|\<|\,|\"|\*|\||\[|\]|\^/", "_", $filename);`

I added these characters to the list of illegal characters:

: [ ] ^ 

Let me know if it works for you. It seems to work for me, but I don't have an exFAT or FAT32 filesystem to really test with.

**Comment by thawkins** [comment]
I will try it out tonight, my version is a little out of date, but I'm
rehosting it on a new machine, it's time it was updated. I'm moving my
torrentwatch/transmission/plex setup from one linux server to another.
Finaly got the daemon version of transmission to play nice.

I also found a way of forcing an ext4 filesystem to write all files with
effective 0774 permission bits by setting the mount options on the fstab
mount to fmask=003,dmask=003, that gives me the ability to have a group of
users/processes (me, plex, transmission, custom scripts) that can write
whilst allowing everybody else to read the media, which is what I was
trying to achieve using the exfat filesystem,  permissions can get to be a
pain in the ass.

**Comment by dchang0** [comment]
Re: "my version is little out of date."
Changing that one line should be possible on even older versions of torrentwatch-xa. That line has been in there for quite a few versions now, and the change is so small that it will not cause trouble. In other words, you don't have to do a full upgrade to the latest version to test out the new line. As long as you find the original line, you can swap it out and try it out.

Yeah, I struggled with transmission for a while myself; had to change the umask within transmission's config file.

Nice work on the ext4 permissions find!

**Comment by dchang0** [comment]
I've tested the new sanitizeFilename() function for a few days, and it's been pushed to the master.

**Comment by thawkins** [comment]
Cool, I may as well do an update to latest version since I am moving
systems.

**Comment by dchang0** [comment]
Let me know how it goes for you.

BTW, while the new sanitizeFilename() function can and does rename the .torrent or .magnet file so that it doesn't have illegal characters in the name, I have noticed that the torrent itself downloaded from the .torrent or .magnet file will usually have the illegal characters, and it's up to the bittorrent client to rename those.

Thus, you might still have trouble with a file that has a colon in the name because the contents of the torrent will probably have the colon in their filename(s) too. So, you will probably still have to deal with the exFAT problem with colons, but outside the reach of torrentwatch-xa.

**Comment by thawkins** [comment]
I just noticed some paths have double slashes in them is. '//' I will try
and work out what is causing it. It's possibly and problem with the
download path.

**Comment by dchang0** [comment]
Double-slashes get treated as a single slash, so even though it's ugly, it should not affect your downloads.

**Comment by thawkins** [comment]
After applying the patch I just downloaded the two series that caused it to
barf before, so it looks like it's a winner.

**Comment by dchang0** [comment]
Excellent--glad to hear it! I'm closing this issue.
